src/rez/build_process.py

```diff
@@ -223,9 +223,26 @@ def get_package_install_path(self, path):
         )
 
     def create_build_context(self, variant, build_type, build_path):
-        """Create a context to build the variant within."""
-        request = variant.get_requires(build_requires=True,
-                                       private_build_requires=True)
+        """Create a context to build the variant within.
+        """
+        # Re-evaluate the variant, so that variables such as 'building' and
+        # 'build_variant_index' are set, and any early-bound package attribs
+        # are re-evaluated wrt these vars. This is done so that attribs such as
+        # 'requires' can change depending on whether a build is occurring or not.
+        #
+        # Note that this re-evaluated variant is ONLY used here, for the purposes
+        # of creating the build context. The variant that is actually installed
+        # is the one evaluated where 'building' is False.
+        #
+        re_evaluated_package = variant.parent.get_reevaluated({
+            "building": True,
+            "build_variant_index": variant.index or 0,
+            "build_variant_requires": variant.variant_requires
+        })
+        re_evaluated_variant = re_evaluated_package.get_variant(variant.index)
```

Comment:
Moving this package and variant re-evaluation into build-process base class `create_build_context` seems better and post-context-resolve re-evaluation can also happen here as well. Pre- and post-re-evaluations are in the same code block now.